src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -38,28 +38,142 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
         /// <remarks>This needs to be called from a background thread. It may hang on the UI thread.</remarks>
         public async Task PerformActionAsync(INuGetUI uiService, UserAction userAction, DependencyObject windowOwner, CancellationToken token)
         {
-            try
+            await PerformActionImplAsync(
+                uiService,
+                () =>
+                {
+                    var projects = uiService.Projects;
+
+                    // TODO: should stable packages allow prerelease dependencies if include prerelease was checked?
+                    // Allow prerelease packages only if the target is prerelease
+                    var includePrelease = userAction.PackageIdentity.Version.IsPrerelease || userAction.Action == NuGetProjectActionType.Uninstall;
+                    var includeUnlisted = userAction.Action == NuGetProjectActionType.Uninstall;
+
+                    var resolutionContext = new ResolutionContext(uiService.DependencyBehavior, includePrelease, includeUnlisted, VersionConstraints.None);
+
+                    return GetActionsAsync(
+                        uiService,
+                        projects,
+                        userAction,
+                        uiService.RemoveDependencies,
+                        uiService.ForceRemove,
+                        resolutionContext,
+                        projectContext: uiService.ProgressWindow,
+                        token: token);
+                },
+                (actions) =>
+                {
+                    return ExecuteActionsAsync(actions, uiService.ProgressWindow, userAction, token);
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Perform the multi-package update action.
+        /// </summary>       
+        /// <remarks>This needs to be called from a background thread. It may hang on the UI thread.</remarks>
+        public async Task PerformUpdateAsync(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
```

Comment:
Unfortunately no, because the API created by John takes List<> as the input. We should ask John to change the API to take IEnumerable<>.